File: src/packs/loader.ts
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { access, readdir, readFile } from 'node:fs/promises';
import { buildPackSummary, normalizeManifest, type SitePackManifest, type SitePackSummary } from './manifest.js';
import { parseInstructions, type SitePackInstructions } from './instructions.js';
import { parseHints, type SitePackHints } from './hints.js';

export interface SitePack {
  rootDir: string;
  manifest: SitePackManifest;
  instructions: SitePackInstructions;
  hints: SitePackHints;
}

export interface LoadedSitePack {
  pack: SitePack;
  summary: SitePackSummary;
  instructionsSummary: string[];
  knownSignals: string[];
}

export async function getDefaultSitePacksRoot(): Promise<string> {
  const here = path.dirname(fileURLToPath(import.meta.url));
  let current = here;

  while (true) {
    const candidate = path.join(current, 'site-packs');
    try {
      await access(candidate);
      return candidate;
    } catch {
      const parent = path.dirname(current);
      if (parent === current) {
        throw new Error(`Unable to locate site-packs directory from ${here}`);
      }
      current = parent;
    }
  }
}

export async function loadSitePack(rootDir: string): Promise<SitePack> {
  const [manifestRaw, instructionsRaw, hintsRaw] = await Promise.all([
    readFile(path.join(rootDir, 'manifest.json'), 'utf8'),
    readFile(path.join(rootDir, 'instructions.md'), 'utf8'),
    readFile(path.join(rootDir, 'hints.json'), 'utf8')
  ]);

  return {
    rootDir,
    manifest: normalizeManifest(JSON.parse(manifestRaw) as unknown),
    instructions: parseInstructions(instructionsRaw),
    hints: parseHints(JSON.parse(hintsRaw) as unknown)
  };
}

export async function loadAllSitePacks(sitePacksRoot?: string): Promise<SitePack[]> {
  const resolvedRoot = sitePacksRoot ?? (await getDefaultSitePacksRoot());
  const entries = await readdir(resolvedRoot, { withFileTypes: true });
  const dirs = entries.filter((entry) => entry.isDirectory()).map((entry) => path.join(resolvedRoot, entry.name));
  return Promise.all(dirs.map((dir) => loadSitePack(dir)));
}

export async function matchSitePackByUrl(url: string, sitePacksRoot?: string): Promise<LoadedSitePack | null> {
  const hostname = new URL(url).hostname.toLowerCase();
  const packs = await loadAllSitePacks(sitePacksRoot);
  const matched = packs.find((pack) =>
    pack.manifest.domains.some((domain) => hostname === domain.toLowerCase() || hostname.endsWith(`.${domain.toLowerCase()}`))
  );

  if (!matched) {
    return null;
  }

  const matchedDomain = matched.manifest.domains.find(
    (domain) => hostname === domain.toLowerCase() || hostname.endsWith(`.${domain.toLowerCase()}`)
  ) ?? matched.manifest.domains[0] ?? hostname;

  return {
    pack: matched,
    summary: buildPackSummary(matched.manifest, matchedDomain),
    instructionsSummary: matched.instructions.summary,
    knownSignals: matched.hints.knownSignals
  };
}

File: browser_platform/packs/loader.py
from __future__ import annotations
import json
from pathlib import Path
from urllib.parse import urlparse

from .manifest import normalize_manifest, build_pack_summary
from .instructions import parse_instructions
from .hints import parse_hints


def _get_default_site_packs_root() -> Path:
    here = Path(__file__).parent
    current = here
    while True:
        candidate = current / 'site-packs'
        if candidate.exists():
            return candidate
        parent = current.parent
        if parent == current:
            raise RuntimeError(f'Unable to locate site-packs directory from {here}')
        current = parent


def load_site_pack(root_dir: Path | str) -> dict:
    root_dir = Path(root_dir)
    manifest_raw = json.loads((root_dir / 'manifest.json').read_text('utf-8'))
    instructions_raw = (root_dir / 'instructions.md').read_text('utf-8')
    hints_raw = json.loads((root_dir / 'hints.json').read_text('utf-8'))
    return {
        'rootDir': str(root_dir),
        'manifest': normalize_manifest(manifest_raw),
        'instructions': parse_instructions(instructions_raw),
        'hints': parse_hints(hints_raw),
    }


def load_all_site_packs(site_packs_root: Path | str | None = None) -> list[dict]:
    resolved_root = Path(site_packs_root) if site_packs_root else _get_default_site_packs_root()
    dirs = [entry for entry in resolved_root.iterdir() if entry.is_dir()]
    return [load_site_pack(d) for d in dirs]


def match_site_pack_by_url(url: str, site_packs_root: Path | str | None = None) -> dict | None:
    hostname = urlparse(url).hostname.lower() if urlparse(url).hostname else ''
    packs = load_all_site_packs(site_packs_root)

    matched = next(
        (
            pack for pack in packs
            if any(
                hostname == d.lower() or hostname.endswith(f'.{d.lower()}')
                for d in pack['manifest']['domains']
            )
        ),
        None,
    )
    if not matched:
        return None

    matched_domain = next(
        (
            d for d in matched['manifest']['domains']
            if hostname == d.lower() or hostname.endswith(f'.{d.lower()}')
        ),
        matched['manifest']['domains'][0] if matched['manifest']['domains'] else hostname,
    )

    return {
        'pack': matched,
        'summary': build_pack_summary(matched['manifest'], matched_domain),
        'instructionsSummary': matched['instructions']['summary'],
        'knownSignals': matched['hints']['knownSignals'],
    }
